### DISCUSSION TITLE
Chain 1/32scan with 1/16 scan panel

### DISCUSSION THREAD

**Discussion by D34TH5TR1K3** [mrcodetastic/ESP32-HUB75-MatrixPanel-DMA#299] [Q&A] [answered]
@mrfaptastic I really love this library. 
Is there a way for me to chain a 64x64 and a 64x32 panel that I have laying around together? I would like to make a 96x64 panel.

I think that they both have two lines active at one time just that the lines on the 1/32 scan move down 32 times to light up everything instead of 16 times. I guess it would take some software modification to use additional bits that "come out" of the 64x64 panel to drive a 64x32.

I'm sure that I'm not the first person to think about that but google did not yield any results so my hopes are not too high although it sounds doable in theory to me.

Both panels are using FM6124.

I thought the 'E' pin on the 1/16 matrix should be shorted to ground since it is not needed but multimeter diode test does not give me a beep and 'E' to GND measures like 7k ohms resistance... not sure if it's a good idea to connect them with a ribbon cable.

Would be nice to have some expert opinion because I would really like to chain them :)
Thank you for reading.

**Comment by mrcodetastic** [comment]
Just simply experiment with setting up the configuration to 2x64px panels and use the virtual display library.

See what happens. I think that should work.

**Reply by D34TH5TR1K3** [reply]
Thank you for the quick reply. I did two experiments:

Setup:
64x64 connected to power and ESP32
64x32 connected to power and output of 64x64 via ribbon cable
1.
Uploaded `1_SimpleTestShapes.ino` (modifications: `#define PANEL_RES_Y 64` `mxconfig.driver = HUB75_I2S_CFG::FM6124;`)

```
void loop() {
  for(int y = 0; y < 64; y++) {
    dma_display->drawPixel(0, y, dma_display->color444(15, 0, 0));
    dma_display->drawPixel(1, y, dma_display->color444(0, 15, 0));
    dma_display->drawPixel(2, y, dma_display->color444(0, 0, 15));
    delay(200);
  }
}
```

If I use code without any chaining, the first matrix will obviously still output the bits that get "pushed out" of the registers into the next matrix. Did this test to see how the 64x32 interprets that.
Result: https://streamable.com/8axun2

2.
Uploaded `ChainedPanels.ino`  
(modifications: `mxconfig.driver = HUB75_I2S_CFG::FM6124;` `#define PANEL_RES_Y 64` `#define NUM_ROWS 2` `#define NUM_COLS 1` `mxconfig.gpio.e = 18;`)
I just left `drawDisplayTest()` and the drawing of the diagonal line uncommented of the drawing functions because it was too cluttered otherwise...
Using serpent:
![IMG20220710210312](https://user-images.githubusercontent.com/77392455/178158532-1463438d-35d3-42b2-addf-f559b3a55238.jpg)
a. colors not matching again
b. 64x32 gets drawn twice like in test #1
c. weird glitching as you can see. Some pixels flicker when I touch the matrices.
Both matrices share the same power (GND should be also connected over the ribbon cable)

I'm not too sure what to make of that but maybe you have an Idea since you have probably debugged these panels a lot of times. I hopy my tests show the issue well. Thank you for taking your time.

EDIT:
Does look like it could work, though. I just have to somehow modify the code to solve the color change which should be easy (if pixel on display 2 swap green and blue [lol]). Also looks like the double drawing that happens because of the different scan rates can be solved via software. I don't know why it's glitchy. The first example with the patterns runs fine without pixel errors on just a single 64x64. What confuses me is that pin E on my 64x32 is neither connected to GND nor not connected but has a resistance to GND, although the 1/16 scan 64x32 does not need it... Maybe I should cut pin E at the ribbon cable?

**Comment by board707** [comment]
I would bet that it won't work. It is not possible to run matrices of different types in one connection.
To run two 64x32 matrices as one 64x64, you have to connect pin E to some kind of switch that will send the top half picture to the first matrix and the bottom half to the second, how it happens on 64x64 matrix

In your case when you connect the output from the 64x64 matrix to the input of the 64x32, the signals from the upper and lower half are mixed, since there is no E pin on the 64x32 matrix to separate these signals.

**Reply by D34TH5TR1K3** [reply]
That makes sense. I thought that I can maybe programm it to send off pixels on the first run or the same pattern twice on the 64x32 but I did not manage to do it.

**Comment by D34TH5TR1K3** [comment]
Appears to be working. Will write up what I did tomorrow since I'm almost asleep now haha.
https://user-images.githubusercontent.com/77392455/178170554-d090675a-d4fa-4c26-ae6f-c1d4b73f05d6.mp4

**Reply by mrcodetastic** [reply]
Wow.  Great work.  I'll be curious to know when you had to do. 

Surely it required some hacking of the core library?

**Comment by board707** [comment]
My congratulations.
Very interesting how it's done

**Comment by D34TH5TR1K3** [comment, marked as answer]
(not perfect) solution:
ESP32 connected to 64x64 .
Output of the 64x64 connected to input of the 64x32 via ribbon cable.

My experiments with the virtualdisplay library were not very successful as you can see in the image I posted. This made me go back to the roots. I used 1_SimpleTestShapes.ino as a base that I modified.
If you just configure ...
```
#define PANEL_RES_X 64
#define PANEL_RES_Y 64
#define PANEL_CHAIN 1
```
... and display a simple animation you will get what you can see in the video that I have already posted:
https://streamable.com/8axun2 (In the video I used PANEL_CHAIN 1 so the second panel basically uses the rgb data that was on the first panel on the frame before since the bits get shifted out.)
[Guide on how these panels work in case you don't understand how the bits are shifted](https://www.bigmessowires.com/2018/05/24/64-x-32-led-matrix-programming/)

Configure `#define PANEL_CHAIN 2` and you have basically the same but you can choose different content for each display since the second one now gets dedicated rgb data. 

Now I had to manipulate the content on my second display (64x32) in three ways to get the desired output:
1. Rotate the image on the second display so then can be placed next to each other nicely
2. Delete some of the rows since they get drawn two times (can be observed in the video if you look closely)
3. Swap green and blue colors so the panels are matching (I did this to the 64x64 matrix and not the 64x32 so the color values that I put in the draw functions would be in the right order R, G, B)

I modified the `MatrixPanel_I2S_DMA::updateMatrixDMABuffer(int16_t x_coord, int16_t y_coord, uint8_t red, uint8_t green, uint8_t blue)` function in `ESP32-HUB75-MatrixPanel-I2S-DMA.cpp` to do these pixel manipulations. I assume that it is a dumb idea to stuff my math functions into this function since it says  
> this is the main method used to update DMA buffer on pixel-by-pixel level so it must be fast, real fast!
above the function but that is what I came up with at 3am. Could also do the pixel manipulation somewhere else I guess. Maybe in the main code. 

Now what did I add to this function? 

```
if(x_coord <= 63 && x_coord >= 0 && y_coord >= 0){
    //Rotate pixels on the second matrix
    //This makes the next code confusing since it swaps x and y
    int yTmp = y_coord;
    y_coord = x_coord;
    x_coord = 63 - yTmp;
    //If pixel is on the part of the virtual matrix that does not exist in reality, 
    //because I am using 64x32 instead of 64x64, dont even draw that pixel.
    if(y_coord <= 31 || y_coord >= 64 ||x_coord >= 64){
      return;
    }
    //Move some pixels so the don't get drawn over each other
    else if(y_coord >= 32 && y_coord < 48){
      y_coord = y_coord - 16;
    }
  }else{
    //swap colors if pixel is on the 64x64
    uint8_t greenTmp = green;
    green = blue;
    blue = greenTmp;
  }
```
Code was written at 3am. It is basically a proof of concept. I had some visual bugs when i wrote it but can't reproduce them so maybe I have fixed it not sure (must have beed a mathematical error)

Code (I know I should do this in Github but everything is a mess so for the time being have this):
[ESP32-HUB75-MatrixPanel-I2S-DMA.cpp](https://pastebin.com/ZXVEkNMm)
[main.cpp](https://pastebin.com/9nqQdSqQ)

**Reply by mrcodetastic** [reply]
Nice work.  Nothing wrong with how you have approached it.  Simple fact is the library isn't designed for chaining different types of panels,  and for whatever reason your 1/16 scan panels has the red and green input the "wrong" way around. 

Glad you got it to work.

**Reply by D34TH5TR1K3** [reply]
Thank you for the support! If you say it is alright I will leave it in this function. I just realized that I can't move the calculations to main since I do not have the individual pixels in main if I draw a circle for example. 
I am so happy that I got it working because I like the aspect ratio and these two displays are just what I already have.
